flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/FunctionService.java

```diff
@@ -0,0 +1,184 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Service for creating configured instances of {@link UserDefinedFunction} using a
+ * {@link FunctionDescriptor}.
+ */
+public class FunctionService {
+
+	private static final Logger LOG = LoggerFactory.getLogger(FunctionService.class);
+
+	/**
+	 * Creates a user-defined function with the given properties and the current thread's
+	 * context class loader.
+	 *
+	 * @param descriptor the descriptor that describes a function
+	 * @return the generated user-defined function
+	 */
+	public static UserDefinedFunction createFunction(FunctionDescriptor descriptor) {
+		return createFunction(descriptor, Thread.currentThread().getContextClassLoader());
+	}
+
+	/**
+	 * Creates a user-defined function with the given properties.
+	 *
+	 * @param descriptor the descriptor that describes a function
+	 * @param classLoader the class loader to load the function and its parameter's classes
+	 * @return the generated user-defined function
+	 */
+	public static UserDefinedFunction createFunction(
+			FunctionDescriptor descriptor,
+			ClassLoader classLoader) {
+		return createFunction(descriptor, classLoader, true);
+	}
+
+	/**
+	 * Creates a user-defined function with the given properties.
+	 *
+	 * @param descriptor the descriptor that describes a function
+	 * @param classLoader the class loader to load the function and its parameter's classes
+	 * @param performValidation whether or not the descriptor should be validated
+	 * @return the generated user-defined function
+	 */
+	public static UserDefinedFunction createFunction(
+			FunctionDescriptor descriptor,
+			ClassLoader classLoader,
+			boolean performValidation) {
+
+		DescriptorProperties properties = new DescriptorProperties(true);
+		properties.putProperties(descriptor.toProperties());
+
+		// validate
+		if (performValidation) {
+			new FunctionDescriptorValidator().validate(properties);
+		}
+
+		// instantiate
+		Tuple2<Class<Object>, Object> tuple2 = generateInstance(
+				HierarchyDescriptorValidator.EMPTY_PREFIX,
+				properties,
+				classLoader);
+
+		if (!UserDefinedFunction.class.isAssignableFrom(tuple2.f0)) {
+			throw new ValidationException(
+					String.format("Instantiated class '%s' is not a user-defined function.", tuple2.f0.getName()));
+		}
+		return (UserDefinedFunction) tuple2.f1;
+	}
+
+	/**
+	 * Recursively generate an instance of a class according the given properties.
+	 *
+	 * @param keyPrefix the prefix to fetch properties
+	 * @param descriptorProperties the descriptor properties that contains the class type information
+	 * @param classLoader the class loader to load the class
+	 * @param <T> type fo the generated instance
+	 * @return an instance of the class
+	 */
+	private static <T> Tuple2<Class<T>, T> generateInstance(
```

Comment:
Do we need to return the class separately? Can't we just call the `.getClass` on the instance whenever necessary?

If it is necessary can we at least have a specialized pojo for that rather than returning a Pojo? But I really believer we can just return the instance.

Reply:
I think you are right. Sometimes we need another class field to deal with null and parent class. But here, looks like class must equal to `.getClass()`.